shared-libs/pouchdb-http-auth-session/src/index.js

```diff
@@ -0,0 +1,134 @@
+const util = require('node:util');
+const { Headers } = require('node-fetch');
+
+const sessionCookieName = 'AuthSession';
+const cookieRegex = new RegExp(`${sessionCookieName}=(.*)`);
+const sessions = {};
+
+const parseCookie = (response) => {
+  const cookie = response.headers.get('set-cookie');
+  if (!cookie) {
+    return;
+  }
+
+  // Unfortunately, node-fetch doesn't handle multiple cookies
+  // https://github.com/node-fetch/node-fetch/issues/251
+  // AuthSession=YWRtaW46NjU5RDRFMzI6Qi5U7t5gHQMn4MgOYkEAX2qH5HZUpn6nKdX8Ik7gDpY; Version=1; Expires=Wed,
+  // 08-Jan-2025 13:46:26 GMT; Max-Age=31536000; Path=/; HttpOnly
+
+  const matches = cookie.match(cookieRegex);
+  if (!matches) {
+    return;
+  }
+
+  const parts = matches[1].split(';').map(item => item.trim().split('='));
+  const session = {
+    token: parts[0][0],
+    expires: new Date(parts.find(part => part[0] === 'Expires')[1]).valueOf(),
+  };
+
+  return session;
+};
+
+const getExistingSession = (db) => {
+  const url = new URL(db.name);
+  url.pathname = '/_session';
+  const urlString = url.toString();
+  if (sessions[urlString] && !isExpired(sessions[urlString])) {
+    return sessions[urlString];
+  }
+};
+
+const getSessionUrl = (db) => {
+  const url = new URL(db.name);
+  url.pathname = '/_session';
+  return url.toString();
+};
+
+const authenticate = async (db) => {
+  const url = getSessionUrl(db);
+
+  const authString = `${db.credentials.username}:${db.credentials.password}`;
+  const token = btoa(decodeURIComponent(encodeURIComponent(authString)));
+  const headers = new Headers();
+  headers.set('Authorization', 'Basic ' + token);
+  headers.set('Content-Type', 'application/json');
+
+  const body = JSON.stringify({ name: db.credentials.username, password: db.credentials.password});
+
+  const response = await db.originalFetch(url.toString(), { method: 'POST', headers, body });
+  updateSession(db, response);
+};
+
+const updateSession = (db, response) => {
+  const session = parseCookie(response);
+  if (session) {
+    const url = getSessionUrl(db);
+    sessions[url] = session;
+  }
+};
+
+const invalidateSession = db => {
+  const url = getSessionUrl(db);
+  delete sessions[url];
+};
+
+const isExpired = (session) => Date.now() > session.expires;
+
+const extractAuth = (opts) => {
+  const url = new URL(opts.name);
+  if (!url.username) {
+    return;
+  }
+
+  opts.credentials = {
+    username: url.username,
+    password: url.password
+  };
+
+  url.username = url.password = '';
```

Comment:
I don't think this is necessary. In my testing, couchdb uses the session token and falls back to basic auth if the session is expired. If you just leave the basic auth there then the session will be automatically established which greatly simplifies the code because you don't need to:

1. Make a _session request - whatever URL the client uses will just work, saving an additional request to the server.
2. Track session expiry - when it expires a new one will be automatically established using basic auth.
3. Invalidate the session - sending basic auth and an expired cookie will just work.
4. Catch the 401 from the server - if it doesn't work with session cookie and basic auth it's not going to work, just return the error to the client.

Reply:
I thought about going this route, but I've been reluctant because this is empirical at best, and there's no documentation around this. I'd rather be comprehensive than assumptive.

Reply:
It's the logical implementation too. I'd rather not mess with the URL if we can avoid it.